File: 2024/day01.cpp
// https://adventofcode.com/2024/day/1

#include "../common/common.h"
#include "../common/rust.h"

#include <print>
#include <ranges>

struct Input {
  Vec<i32> left;
  Vec<i32> right;
};

auto parse(String const& filename) -> Input {
  auto input = Input{};
  for (str line : aoc::views::read_lines(filename)) {
    let parsed = aoc::split_once<i32>(line, "   ");
    input.left.push_back(parsed[0]);
    input.right.push_back(parsed[1]);
  }
  return input;
}

fn sum_diffs(Vec<i32> const& left_list, Vec<i32> const& right_list) -> i32 {
  AOC_ASSERT(left_list.size() == right_list.size(), "Lists must be same size");
  let left = aoc::ranges::sorted(left_list);
  let right = aoc::ranges::sorted(right_list);

  return aoc::ranges::accumulate(
      stdv::zip_transform([](i32 l, i32 r) { return aoc::abs(r - l); }, //
                          left, right),
      0);
}

static_assert(sum_diffs({2}, {7}) == 5);
static_assert(sum_diffs({3, 4, 2, 1, 3, 3}, {4, 3, 5, 3, 9, 3}) == 11);

fn similarity_score(Vec<i32> const& left_list, Vec<i32> const& right_list)
    -> i32 {
  auto right_counts = aoc::flat_map<i32, i32>{};
  for (auto value : right_list) {
    ++right_counts[value];
  }

  return aoc::ranges::accumulate(
      left_list | stdv::transform([&](i32 value) {
        let it = right_counts.find(value);
        return value * (it == right_counts.end() ? 0 : it->second);
      }),
      0);
}

static_assert(similarity_score({3}, {4, 3, 5, 3, 9, 3}) == 9);
static_assert(similarity_score({3, 3}, {4, 3, 5, 3, 9, 3}) == 18);
static_assert(similarity_score({3, 4}, {4, 3, 5, 3, 9, 3}) == 13);
static_assert(similarity_score({3, 4, 3}, {4, 3, 5, 3, 9, 3}) == 22);
static_assert(similarity_score({3, 4, 2}, {4, 3, 5, 3, 9, 3}) == 13);
static_assert(similarity_score({3, 4, 2, 1}, {4, 3, 5, 3, 9, 3}) == 13);
static_assert(similarity_score({3, 4, 2, 1, 3}, {4, 3, 5, 3, 9, 3}) == 22);
static_assert(similarity_score({3, 4, 2, 1, 3, 3}, {4, 3, 5, 3, 9, 3}) == 31);

int main() {
  std::println("Part 1");
  let example = parse("day01.example");
  AOC_EXPECT_RESULT(11, sum_diffs(example.left, example.right));
  let input = parse("day01.input");
  AOC_EXPECT_RESULT(3569916, sum_diffs(input.left, input.right));

  std::println("Part 2");
  AOC_EXPECT_RESULT(31, similarity_score(example.left, example.right));
  AOC_EXPECT_RESULT(26407426, similarity_score(input.left, input.right));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2024/src/day01.rs
use rustc_hash::FxHashMap;

struct Input {
    left: Vec<i32>,
    right: Vec<i32>,
}

fn parse(filename: &str) -> Input {
    let (left, right) = aoc::file::read_lines(filename)
        .iter()
        .map(|line| {
            let mut numbers = line.split_whitespace();
            let left = numbers.next().unwrap().parse::<i32>().unwrap();
            let right = numbers.next().unwrap().parse::<i32>().unwrap();
            (left, right)
        })
        .unzip();
    Input { left, right }
}

fn solve_case1(input: &Input) -> i32 {
    let mut left_list = input.left.clone();
    let mut right_list = input.right.clone();
    left_list.sort_unstable();
    right_list.sort_unstable();

    left_list
        .iter()
        .zip(right_list.iter())
        .map(|(&left, &right)| (left - right).abs())
        .sum()
}

fn solve_case2(input: &Input) -> i32 {
    let mut right_counts = FxHashMap::default();
    for &value in &input.right {
        *right_counts.entry(value).or_insert(0) += 1;
    }

    input
        .left
        .iter()
        .map(|value| value * right_counts.get(value).copied().unwrap_or(0))
        .sum()
}

fn main() {
    println!("Part 1");
    let example = parse("day01.example");
    aoc::expect_result!(11, solve_case1(&example));
    let input = parse("day01.input");
    aoc::expect_result!(3569916, solve_case1(&input));

    println!("Part 2");
    aoc::expect_result!(31, solve_case2(&example));
    aoc::expect_result!(26407426, solve_case2(&input));
}
